src/realm/backup_restore.cpp

```diff
@@ -0,0 +1,172 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ **************************************************************************/
+
+#include <realm/backup_restore.hpp>
+#include <realm/util/file.hpp>
+
+#include <vector>
+#include <chrono>
+
+namespace realm {
+
+/*
+ * IMPORTANT: The following two arrays must be kept updated
+ * as new versions are released or if rollback is ever done.
+ */
+
+using version_list_t = BackupHandler::version_list_t;
+using version_time_list_t = BackupHandler::version_time_list_t;
+
+// Note: accepted versions should have new versions added at front
+const version_list_t accepted_versions_{20, 11, 10, 9, 8, 7, 6, 0};
+
+// the pair is <version, age-in-seconds>
+// we keep backup files in 3 months: 3*30*24*60*60 secs
+constexpr int three_months = 3 * 30 * 24 * 60 * 60;
+const version_time_list_t delete_versions_{{20, three_months}, {11, three_months}, {10, three_months},
+                                           {9, three_months},  {8, three_months},  {7, three_months},
+                                           {6, three_months}};
+
+version_list_t accepted_versions{accepted_versions_};
+version_time_list_t delete_versions{delete_versions_};
+
+
+// helper functions
+std::string backup_name(std::string prefix, int version)
+{
+    return prefix + "v" + std::to_string(version) + ".backup.realm";
+}
+
+bool backup_exists(std::string prefix, int version)
+{
+    std::string fname = backup_name(prefix, version);
+    return util::File::exists(fname);
+}
+
+void BackupHandler::fake_versions(const version_list_t& accepted, const version_time_list_t& to_delete)
+{
+    accepted_versions = accepted;
+    delete_versions = to_delete;
+}
+
+void BackupHandler::unfake_versions()
+{
+    accepted_versions = accepted_versions_;
+    delete_versions = delete_versions_;
+}
+
+std::string BackupHandler::get_prefix_from_path(std::string path)
+{
+    // prefix is everything but the suffix here, so start from the back
+    for (auto i = path.size() - 1; i; --i) {
+        if (path[i] == '.')
+            return path.substr(0, i + 1);
+    }
+    // if not on normal "prefix.suffix" form add "."
+    return path + ".";
+}
+
+BackupHandler::BackupHandler(const std::string& path)
+{
+    m_path = path;
+    m_prefix = get_prefix_from_path(path);
+}
+
+bool BackupHandler::must_restore_from_backup(int current_file_format_version)
+{
+    if (current_file_format_version == 0)
+        return false;
+    for (auto i : accepted_versions) {
```

Comment:
s/i/v/ (or version)

`i` is conventionally used as an index, so using it as the value of the iteration is misleading


ditto for all of these loops